components/camera/videosource/query.go

```diff
@@ -0,0 +1,225 @@
+
+// GetNamedVideoSource attempts to find a device (not a screen) by the given name.
+// If name is empty, it finds any device.
+func getReaderAndDriver(
+	name string,
+	constraints mediadevices.MediaStreamConstraints,
+	logger logging.Logger,
+) (video.Reader, driver.Driver, error) {
+	var ptr *string
+	if name == "" {
+		ptr = nil
+	} else {
+		ptr = &name
+	}
+	d, selectedMedia, err := getUserVideoDriver(constraints, ptr, logger)
+	if err != nil {
+		return nil, nil, err
+	}
+	reader, err := newReaderFromDriver(d, selectedMedia)
+	if err != nil {
+		return nil, nil, err
+	}
+	return reader, d, nil
+}
+
+func getUserVideoDriver(
+	constraints mediadevices.MediaStreamConstraints,
+	label *string,
+	logger logging.Logger,
+) (driver.Driver, prop.Media, error) {
+	var videoConstraints mediadevices.MediaTrackConstraints
+	if constraints.Video != nil {
+		constraints.Video(&videoConstraints)
+	}
+	return selectVideo(videoConstraints, label, logger)
+}
+
+func newReaderFromDriver(
+	videoDriver driver.Driver,
+	mediaProp prop.Media,
+) (video.Reader, error) {
+	recorder, ok := videoDriver.(driver.VideoRecorder)
+	if !ok {
+		return nil, errors.New("driver not a driver.VideoRecorder")
+	}
+
+	if ok, err := driver.IsAvailable(videoDriver); !errors.Is(err, availability.ErrUnimplemented) && !ok {
+		return nil, errors.Wrap(err, "video driver not available")
+	} else if driverStatus := videoDriver.Status(); driverStatus != driver.StateClosed {
+		return nil, errors.New("video driver in use")
+	} else if err := videoDriver.Open(); err != nil {
+		return nil, errors.Wrap(err, "cannot open video driver")
+	}
+
+	mediaProp.DiscardFramesOlderThan = time.Second
+	reader, err := recorder.VideoRecord(mediaProp)
+	if err != nil {
+		return nil, err
+	}
+	return reader, nil
+}
+
+func labelFilter(target string, useSep bool) driver.FilterFn {
+	return driver.FilterFn(func(d driver.Driver) bool {
+		if !useSep {
+			return d.Info().Label == target
+		}
+		labels := strings.Split(d.Info().Label, camera.LabelSeparator)
+		for _, label := range labels {
+			if label == target {
+				return true
+			}
+		}
+		return false
+	})
+}
+
+func selectVideo(
+	constraints mediadevices.MediaTrackConstraints,
+	label *string,
+	logger logging.Logger,
+) (driver.Driver, prop.Media, error) {
+	return selectBestDriver(getVideoFilterBase(), getVideoFilter(label), constraints, logger)
+}
+
+func getVideoFilterBase() driver.FilterFn {
+	typeFilter := driver.FilterVideoRecorder()
+	notScreenFilter := driver.FilterNot(driver.FilterDeviceType(driver.Screen))
+	return driver.FilterAnd(typeFilter, notScreenFilter)
+}
+
+func getVideoFilter(label *string) driver.FilterFn {
+	filter := getVideoFilterBase()
+	if label != nil {
+		filter = driver.FilterAnd(filter, labelFilter(*label, true))
+	}
+	return filter
+}
+
+// select implements SelectSettings algorithm.
+// Reference: https://w3c.github.io/mediacapture-main/#dfn-selectsettings
+func selectBestDriver(
```

Comment:
Does the behavior of the driver matcher here change at all from previous code? Does it accept width, height, and framerate props that are not an exact fit?

Reply:
It should be the same since I copy pasted it from where it used to be referenced in gostream